File: backend/app/schemas/agenda.py
"""
Agenda schemas for meeting agendas
"""
from datetime import datetime
from typing import Optional, List
from pydantic import BaseModel, Field
from uuid import UUID


class AgendaItemBase(BaseModel):
    """Base agenda item schema"""
    order_index: int = Field(ge=0, description="Order of the item in agenda")
    title: str
    duration_minutes: Optional[int] = Field(default=15, ge=1)
    presenter_id: Optional[UUID] = None
    presenter_name: Optional[str] = None
    description: Optional[str] = None
    notes: Optional[str] = None


class AgendaItemCreate(AgendaItemBase):
    """Create agenda item request"""
    pass


class AgendaItemUpdate(BaseModel):
    """Update agenda item request"""
    order_index: Optional[int] = None
    title: Optional[str] = None
    duration_minutes: Optional[int] = None
    presenter_id: Optional[UUID] = None
    presenter_name: Optional[str] = None
    description: Optional[str] = None
    notes: Optional[str] = None
    status: Optional[str] = None  # pending, in_progress, completed


class AgendaItem(AgendaItemBase):
    """Agenda item response"""
    id: UUID
    meeting_id: UUID
    status: str = "pending"
    created_at: datetime
    updated_at: datetime
    
    class Config:
        from_attributes = True


class AgendaList(BaseModel):
    """List of agenda items response"""
    items: List[AgendaItem]
    total: int
    total_duration_minutes: int


class AgendaCreate(BaseModel):
    """Create full agenda request"""
    meeting_id: UUID
    items: List[AgendaItemCreate]


class AgendaGenerateRequest(BaseModel):
    """Request to generate agenda using AI"""
    meeting_id: UUID
    meeting_title: str
    meeting_type: Optional[str] = None
    meeting_description: Optional[str] = None
    duration_minutes: Optional[int] = 60
    participants: Optional[List[str]] = None
    context: Optional[str] = None  # Additional context for AI


class AgendaGenerateResponse(BaseModel):
    """Response from AI agenda generation"""
    items: List[AgendaItemCreate]
    total_duration_minutes: int
    ai_notes: Optional[str] = None
    is_saved: bool = False


class AgendaSaveRequest(BaseModel):
    """Request to save generated agenda"""
    meeting_id: UUID
    items: List[AgendaItemCreate]

File: electron/src/renderer/lib/api/agenda.ts
// ============================================
// AGENDA API
// API functions for agenda operations
// ============================================

import api from '../apiClient';

export interface AgendaItem {
  id: string;
  meeting_id: string;
  order_index: number;
  title: string;
  duration_minutes: number;
  presenter_id?: string;
  presenter_name?: string;
  description?: string;
  notes?: string;
  status: 'pending' | 'in_progress' | 'completed';
  created_at: string;
  updated_at: string;
}

export interface AgendaItemCreate {
  order_index: number;
  title: string;
  duration_minutes?: number;
  presenter_id?: string;
  presenter_name?: string;
  description?: string;
  notes?: string;
}

export interface AgendaItemUpdate {
  order_index?: number;
  title?: string;
  duration_minutes?: number;
  presenter_id?: string;
  presenter_name?: string;
  description?: string;
  notes?: string;
  status?: 'pending' | 'in_progress' | 'completed';
}

export interface AgendaList {
  items: AgendaItem[];
  total: number;
  total_duration_minutes: number;
}

export interface AgendaGenerateRequest {
  meeting_id: string;
  meeting_title: string;
  meeting_type?: string;
  meeting_description?: string;
  duration_minutes?: number;
  participants?: string[];
  context?: string;
}

export interface AgendaGenerateResponse {
  items: AgendaItemCreate[];
  total_duration_minutes: number;
  ai_notes?: string;
  is_saved: boolean;
}

export interface AgendaSaveRequest {
  meeting_id: string;
  items: AgendaItemCreate[];
}

const ENDPOINT = '/agenda';

export const agendaApi = {
  /**
   * List all agenda items for a meeting
   */
  listByMeeting: async (meetingId: string): Promise<AgendaList> => {
    return api.get<AgendaList>(`${ENDPOINT}/meeting/${meetingId}`);
  },

  /**
   * Get a single agenda item by ID
   */
  getItem: async (itemId: string): Promise<AgendaItem> => {
    return api.get<AgendaItem>(`${ENDPOINT}/item/${itemId}`);
  },

  /**
   * Create a new agenda item
   */
  createItem: async (meetingId: string, data: AgendaItemCreate): Promise<AgendaItem> => {
    return api.post<AgendaItem>(`${ENDPOINT}/meeting/${meetingId}/item`, data);
  },

  /**
   * Update an agenda item
   */
  updateItem: async (itemId: string, data: AgendaItemUpdate): Promise<AgendaItem> => {
    return api.put<AgendaItem>(`${ENDPOINT}/item/${itemId}`, data);
  },

  /**
   * Delete an agenda item
   */
  deleteItem: async (itemId: string): Promise<void> => {
    return api.delete<void>(`${ENDPOINT}/item/${itemId}`);
  },

  /**
   * Generate agenda using AI
   */
  generate: async (request: AgendaGenerateRequest): Promise<AgendaGenerateResponse> => {
    return api.post<AgendaGenerateResponse>(`${ENDPOINT}/generate`, request);
  },

  /**
   * Save a complete agenda (replaces existing)
   */
  save: async (data: AgendaSaveRequest): Promise<AgendaList> => {
    return api.post<AgendaList>(`${ENDPOINT}/save`, data);
  },

  /**
   * Reorder agenda items
   */
  reorder: async (meetingId: string, itemIds: string[]): Promise<AgendaList> => {
    return api.post<AgendaList>(`${ENDPOINT}/meeting/${meetingId}/reorder`, itemIds);
  },
};

export default agendaApi;

